File: partiql-plan/src/main/java/org/partiql/plan/rex/RexTable.java
package org.partiql.plan.rex;

import org.jetbrains.annotations.NotNull;
import org.partiql.plan.Operand;
import org.partiql.plan.OperatorVisitor;
import org.partiql.spi.catalog.Table;

import java.util.List;

/**
 * Global variable references e.g. tables and views.
 */
public abstract class RexTable extends RexBase {

    /**
     * @return new RexTable instance
     */
    @NotNull
    public static RexTable create(@NotNull Table table) {
        return new Impl(table);
    }

    /**
     * @return the table implementation.
     */
    public abstract Table getTable();

    @NotNull
    @Override
    protected final RexType type() {
        return RexType.of(getTable().getSchema());
    }

    @NotNull
    @Override
    protected final List<Operand> operands() {
        return List.of();
    }

    @Override
    public <R, C> R accept(OperatorVisitor<R, C> visitor, C ctx) {
        return visitor.visitTable(this, ctx);
    }

    private static class Impl extends RexTable {

        private final Table table;

        private Impl(Table table) {
            this.table = table;
        }

        @Override
        public Table getTable() {
            return table;
        }
    }
}

File: partiql-plan/src/main/kotlin/org/partiql/plan/rex/RexTable.kt
package org.partiql.plan.rex

import org.partiql.spi.catalog.Table

/**
 * Global variable references e.g. tables and views.
 */
public interface RexTable : Rex {

    public fun getTable(): Table

    override fun getChildren(): Collection<Rex> = emptyList()

    override fun <R, C> accept(visitor: RexVisitor<R, C>, ctx: C): R = visitor.visitTable(this, ctx)
}

/**
 * Default [RexTable] implementation.
 */
internal class RexTableImpl(table: Table) : RexTable {

    // DO NOT USE FINAL
    private var _table = table
    private var _type = RexType(table.getSchema())

    override fun getTable(): Table = _table

    override fun getType(): RexType = _type

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is RexTable) return false
        if (_table != other.getTable()) return false
        return true
    }

    override fun hashCode(): Int {
        var result = 1
        result = 31 * result + _table.hashCode()
        return result
    }
}
